### ISSUE TITLE
broker memory optimization

### ISSUE THREAD

**Issue by alka79** [martin-ger/uMQTTBroker#48]
Hi,
I start to use the broker on an ESP8266 that is already hosting a webserver and doing other stuff. 
Install and initial tests were fine but I face memory issues. At some point, the ESP crashed and I could see the error message about lack of memory.

My MQTT needs are quite limited : 2 simultaneous clients max. One client always connected (it is another ESP with PubSubClient) , the other clients are temporary connections from a smartphone, only one a time.
I need only 5 topics overall. The clients may publish to several but only subscribe to 1 topic.
The "local_client" subscribes to everything, but this one does not use extra ram.
The size of the messages published by the clients are small in size. The largest, (topic+json payload) is 50 char long.
The clients subscribe to a single topic where only the local_client publishes to. This topic will receive JSON messages up to 100 char long. 

I could see two optimizations : 

**max_subscriptions and max_retained_topics**
The broker is started with the defaults at 30. 
Retained topics is easy to understand. I changed it to 5.
Max subscriptions is harder to understand. A client can subscribe to # or a single topic. Does it count for one or many subscriptions ? 
When looking at the broker code, I see that "max subscriptions" is actually linked to the number of clients connected. I changed it to 10.
Unfortunately with these changes, the heap remains almost unchanged. 

**MQTT_BUF_SIZE and QUEUE_BUFFER_SIZE**
defined in src/mqtt/defaults.h
_MQTT_BUF_SIZE (default 1024) is the max size of pending inbound messages for one connection
QUEUE_BUFFER_SIZE (default 2048) is the max size of all pending outbound messages for one connection_

At client init, the broker tries to allocate memory of both buffers size (3K bytes). If there is not enough heap left, the connection is rejected. (That is when my ESP crashed.)

I guess that I could lower buffer sizes which would significantly lower the memory used per active client connection.

Two questions : 
- in my situation, how much can I lower both buffer sizes :  Is there a rule? educated guess ? or is it try and error process ?

- The number of clients is exposed by onConnect.  How can I "force reject" a 3rd connection request ?

Thank you

**Comment by martin-ger** [comment]
max_subscriptions and max_retained_topics are just arrays with some meta-infos, small values don't save a lot. Actual subscriptions topic-names and retained topics names and contents are allocated additionally. Each subscribed topic string counts as one subscription per subscribing client, but '#' or other wildcards are just one subscription.

The queue sizes depend on your expected message sizes. If you only have a few bytes from time to time, 512 for both is probably enough.

If you want to block a 3rd connection, just return a "false" from onConnect.

**Comment by alka79** [comment]
thanks :) I 'll give it a try.

**Comment by alka79** [comment]
just to share my findings : 
I noticed that an ESP8266 TCP connection uses about 3KB ram when active. That is just for the core internal stuff. 
The MQTT connection, when active, needs by default 3KB more (MQTT_BUF_SIZE + QUEUE_BUFFER_SIZE ). 
This is all released when the MQTT client deconnects and TCP is closed.

This is per client, and the sketch surely needs additional RAM on top. It is easy to quickly run out of RAM if the number of simultaneous clients grows. The behavior becomes hazardous up to crash the ESP8266.

I have reduced both buffers to 512B. Also reworked my sketch to save RAM here and there. It works fine now with 1 permanent client and 2 more who connect temporarily. Never more than 2 simultaneous clients and it runs smoothly in my situation.
